File: app/src/main/golang/MagneticField.go
package geomag

type MagneticField struct {
	Declination         float64
	DeclinationSV       float64
	Inclination         float64
	InclinationSV       float64
	HorizontalIntensity float64
	HorizontalSV        float64
	NorthComponent      float64
	NorthSV             float64
	EastComponent       float64
	EastSV              float64
	VerticalComponent   float64
	VerticalSV          float64
	TotalIntensity      float64
	TotalSV             float64
}

func newMagneticField(
	declination, declinationSV,
	inclination, inclinationSV,
	horizontalIntensity, horizontalSV,
	northComponent, northSV,
	eastComponent, eastSV,
	verticalComponent, verticalSV,
	totalIntensity, totalSV float64,
) *MagneticField {
	return &MagneticField{
		Declination:         declination,
		DeclinationSV:       declinationSV,
		Inclination:         inclination,
		InclinationSV:       inclinationSV,
		HorizontalIntensity: horizontalIntensity,
		HorizontalSV:        horizontalSV,
		NorthComponent:      northComponent,
		NorthSV:             northSV,
		EastComponent:       eastComponent,
		EastSV:              eastSV,
		VerticalComponent:   verticalComponent,
		VerticalSV:          verticalSV,
		TotalIntensity:      totalIntensity,
		TotalSV:             totalSV,
	}
}

File: app/src/main/kotlin/com/sanmer/geomag/model/MagneticField.kt
package com.sanmer.geomag.model

import go.geomag.MagneticField as GoMagneticField

data class MagneticField(
    val declination: Double,
    val declinationSV: Double,
    val inclination: Double,
    val inclinationSV: Double,
    val horizontalIntensity: Double,
    val horizontalSV: Double,
    val northComponent: Double,
    val northSV: Double,
    val eastComponent: Double,
    val eastSV: Double,
    val verticalComponent: Double,
    val verticalSV: Double,
    val totalIntensity: Double,
    val totalSV: Double
) {
    companion object {
        fun empty() = MagneticField(
            0.0, 0.0,
            0.0, 0.0,
            0.0, 0.0,
            0.0, 0.0,
            0.0, 0.0,
            0.0, 0.0,
            0.0, 0.0
        )
    }
}

fun GoMagneticField.toField() = MagneticField(
    declination = declination,
    declinationSV = declinationSV.orZero(),
    inclination = inclination,
    inclinationSV = inclinationSV.orZero(),
    horizontalIntensity = horizontalIntensity,
    horizontalSV = horizontalSV.orZero(),
    northComponent = northComponent,
    northSV = northSV.orZero(),
    eastComponent = eastComponent,
    eastSV = eastSV.orZero(),
    verticalComponent = verticalComponent,
    verticalSV = verticalSV.orZero(),
    totalIntensity = totalIntensity,
    totalSV = totalSV.orZero()
)

private fun Double.orZero() = if (isNaN()) 0.0 else this
